pypif_sdk/func/calculate_funcs.py

```diff
@@ -0,0 +1,280 @@
+        formula_string = formula_string[0:this_start] + parenth_expanded + formula_string[this_end:]
+        return _expand_formula_(formula_string)
+
+def _expand_hydrate_(hydrate_pos, formula_string):
+    """
+    Handles the expansion of hydrate portions of a chemical formula, and expands out the coefficent to all elements
+
+    :param hydrate_pos is the index in the formula_string of the · symbol
+    :param formula_string is the unexpanded formula string
+    returns a formula string without the · character with the hydrate portion expanded out
+    """
+    hydrate = formula_string[hydrate_pos + 1:]
+    hydrate_string = ""
+    multiplier = float(re.search(r'^[\d\.]+', hydrate).group())
+    element_array = re.findall('[A-Z][^A-Z]*', hydrate)
+    for e in element_array:
+        occurance_array = re.findall('[0-9][^0-9]*', e)
+        if len(occurance_array) == 0:
+            occurance_array.append(1)
+        for o in occurance_array:
+            symbol = re.findall('[A-Z][a-z]*', e)
+            total_num = float(o)*multiplier
+            if total_num.is_integer():
+                    total_num = int(total_num)
+            total_str = str(total_num)
+            if total_str == "1":
+                total_str = ""
+            new_string = symbol[0] + total_str
+            hydrate_string += new_string
+    return formula_string[:hydrate_pos] + hydrate_string
+
+def _create_compositional_array_(expanded_chemical_formaula_string):
+    """
+    Splits an expanded chemical formula string into an array of dictionaries containing information about each element
+    :param expanded_chemical_formaula_string is a clean (not necessarily emperical, but without any special characters) chemical formula string, as returned by _expand_formula_()
+
+    Returns an array of dictionaries
+    """
+    element_array = re.findall('[A-Z][^A-Z]*', expanded_chemical_formaula_string)
+    split_element_array = []
+    for s in element_array:
+        m = re.match(r"([a-zA-Z]+)([0-9\.]*)", s, re.I)
+        if m:
+            items = m.groups()
+            if items[1] == "":
+                items = (items[0], 1)
+        this_e = {"symbol": items[0], "occurances": float(items[1])}
+        
+        split_element_array.append(this_e)
+    return split_element_array
+
+def _consolidate_elemental_array_(elemental_array):
+    """
+    Accounts for non-empirical chemical formulas by taking in the compositional array generated by _create_compositional_array_() and returning a consolidated array of dictionaries with no repeating elements
+
+    :param elemental_array is an elemental array generated from _create_compositional_array_()
+
+    Returns an array of element dictionaries
+    """
+    condensed_array = []
+    for e in elemental_array:
+        exists = False
+        for k in condensed_array:
+            if k["symbol"] == e["symbol"]:
+                exists = True
+                k["occurances"] += e["occurances"]
```

Comment:
I think you want to add a `break` here, in case the same symbol shows up in `condensed_array` multiple times.

Reply:
It should never do that because it always starts with an empty array, but I added the break anyway